src/core/timeouts.ts

```diff
@@ -0,0 +1,57 @@
+import { Context } from '../context'
+import { formatWithOptions } from 'util'
+import Yallist = require('yallist')
+
+interface Drift<C extends Context> {
+  readonly ctx: C
+  readonly timeoutsAt: number
+}
+
+const MIN_TIMEOUT_DURATION = 5_000 // 5s in ms
+
+export class Timeouts<C extends Context> {
+  private readonly list = new Yallist<Drift<C>>()
+
+  add(drift: Drift<C>) {
+    const node = new Yallist.Node(drift)
+    this.list.pushNode(node)
+    this.runTimer()
+    return () => {
+      if (node.list != null) {
+        this.list.removeNode(node)
+      }
+    }
+  }
+
+  handleTimeout = (drift: Drift<C>) => {
+    throw new Error(
+      formatWithOptions(
+        { depth: 2 },
+        'Update processing timed out:',
+        drift.ctx.update
+      )
+    )
+  }
+
+  private isTimerRunning = false
+  private runTimer() {
+    const node = this.list.tail // head works too, sets timers more often
+    if (node == null || this.isTimerRunning) return
+    this.isTimerRunning = true
+    const timeLeft = node.value.timeoutsAt - Date.now()
+    const ms = Math.max(timeLeft, MIN_TIMEOUT_DURATION)
+    setTimeout(() => {
+      try {
+        while (true) {
```

Comment:
1\. Why? 2. Should I use `for` loop?

Reply:
I'll leave as-is for now, I'm changing other things right now

Reply:
You have to be extremely cautious about changing anything, otherwise you may end up with an infinite loop that is very hard to debug. Move the condition to the loop is always safer and easier to read (no need for a flow analysis implemented in brainware).